services/filter/src/filter/routes/filter.py

```diff
@@ -0,0 +1,240 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 The HuggingFace Authors.
+
+import logging
+import os.path
+from typing import Optional
+
+import duckdb
+import pyarrow as pa
+import pyarrow.parquet as pq
+from datasets import Features, Value
+from libapi.authentication import auth_check
+from libapi.exceptions import (
+    ApiError,
+    InvalidParameterError,
+    MissingRequiredParameterError,
+)
+from libapi.utils import (
+    Endpoint,
+    are_valid_parameters,
+    get_json_api_error_response,
+    get_json_ok_response,
+    to_rows_list,
+)
+from libcommon.exceptions import UnexpectedError
+from libcommon.parquet_utils import ParquetFileMetadataItem
+from libcommon.processing_graph import ProcessingGraph
+from libcommon.prometheus import StepProfiler
+from libcommon.simple_cache import get_previous_step_or_raise
+from libcommon.storage import StrPath
+from libcommon.utils import PaginatedResponse
+from libcommon.viewer_utils.features import (
+    get_supported_unsupported_columns,
+    to_features_list,
+)
+from starlette.requests import Request
+from starlette.responses import Response
+
+MAX_ROWS = 100
+
+# audio still has some errors when librosa is imported
+UNSUPPORTED_FEATURES = [Value("binary")]
+
+FILTER_QUERY = """\
+    SELECT {columns}
+    FROM read_parquet({parquet_file_urls})
```

Comment:
we use Duckdb to query parquet files, but we also have the Duckdb version of the file.

1. Would it make sense to use the Duckdb file instead? what are the pros/cons?
2. A specific aspect of parquet vs. Duckdb: some datasets are huge, with many parquet files (might be more than 1TB of parquet files), while the Duckdb index file is always under 10GB at most. How does/filter perform on big datasets?

Reply:
I have no idea of the feasibility of using the DuckDB index instead of the Parquet files. Maybe @AndreaFrancis can give some insights?
- Is it possible to use the DuckDB index to filter the files based on a WHERE clause (like e.g. `age = 30`)?

Reply:
I think we can use the same file as in /search. I think the index does not affect the performance because it creates like another table in another schema.
**Duckdb**
pros:
- We can share the downloaded data for /filter and /search.  
- I think that native SQL queries on an already physical DB should be faster than SQL-parquet

**Using remote parquet**
pros:
- No need to download files 
cons:
- Connectivity issues

Reply:
Also:
- duckdb: limited to the first 5GB
- remote parquet: no limit (see https://huggingface.co/datasets/EleutherAI/the_pile_deduplicated/tree/main/data for example), but... the filter operation might take ages

Reply:
Thanks a lot for your insights, @AndreaFrancis and @severo. :hugs: 

I think I will use the index then, so that /search and /filter are aligned.

Reply:
Note that, if it makes sense, they can share the same "service" (ie: two endpoints: /search and /filter) in services/search. I don't know if it make sense, but it's worth evaluating. cc @AndreaFrancis

Reply:
I think it makes sense, we would have to increase the number of replicas for the existing service

Reply:
Thanks again for your hints. I will need your help with the replicas...

Reply:
re Replicas: it's just the number of pods that run the service. It's set here:

https://github.com/huggingface/datasets-server/blob/1afb2818bf88746494bf2770dfc33616eb68c661/chart/env/prod.yaml#L342

We will possibly have to multiply the number by 2

Reply:
Note that DuckDB indexes are currently only created for datasets that have FTS-indexable (=string) columns. We would have to change this to have a DuckDB file for every datasets

Reply:
Ups, thanks @lhoestq. Should that also be addressed in this PR, or better in another one?

Reply:
Better in another one IMO. For now you can raise an error if the duckdb-index step is missing or has an error

Reply:
OK. Thanks again.

Reply:
> Note that DuckDB indexes are currently only created for datasets that have FTS-indexable (=string) columns. We would have to change this to have a DuckDB file for every datasets

done here: https://github.com/huggingface/datasets-server/pull/1854